receiver/receiver.go

```diff
@@ -23,201 +20,66 @@ var (
-	cfg component.Config,
-	nextConsumer consumer.Metrics,
-) (Metrics, error) {
-	if f == nil {
-		return nil, component.ErrDataTypeIsNotSupported
-	}
-	return f(ctx, set, cfg, nextConsumer)
-}
+type CreateMetricsFunc = internal.CreateMetricsFunc
 
 // CreateLogsFunc is the equivalent of ReceiverFactory.CreateLogsReceiver().
-type CreateLogsFunc func(context.Context, Settings, component.Config, consumer.Logs) (Logs, error)
-
-// CreateLogsReceiver implements Factory.CreateLogsReceiver().
-func (f CreateLogsFunc) CreateLogsReceiver(
-	ctx context.Context,
-	set Settings,
-	cfg component.Config,
-	nextConsumer consumer.Logs,
-) (Logs, error) {
-	if f == nil {
-		return nil, component.ErrDataTypeIsNotSupported
-	}
-	return f(ctx, set, cfg, nextConsumer)
-}
-
-type factory struct {
-	cfgType component.Type
-	component.CreateDefaultConfigFunc
-	CreateTracesFunc
-	tracesStabilityLevel component.StabilityLevel
-	CreateMetricsFunc
-	metricsStabilityLevel component.StabilityLevel
-	CreateLogsFunc
-	logsStabilityLevel component.StabilityLevel
-}
-
-func (f *factory) Type() component.Type {
-	return f.cfgType
-}
-
-func (f *factory) unexportedFactoryFunc() {}
-
-func (f *factory) TracesReceiverStability() component.StabilityLevel {
-	return f.tracesStabilityLevel
-}
-
-func (f *factory) MetricsReceiverStability() component.StabilityLevel {
-	return f.metricsStabilityLevel
-}
-
-func (f *factory) LogsReceiverStability() component.StabilityLevel {
-	return f.logsStabilityLevel
-}
+type CreateLogsFunc = internal.CreateLogsFunc
 
 // WithTraces overrides the default "error not supported" implementation for CreateTracesReceiver and the default "undefined" stability level.
 func WithTraces(createTracesReceiver CreateTracesFunc, sl component.StabilityLevel) FactoryOption {
-	return factoryOptionFunc(func(o *factory) {
-		o.tracesStabilityLevel = sl
-		o.CreateTracesFunc = createTracesReceiver
-	})
+	return internal.WithTraces(createTracesReceiver, sl)
 }
 
 // WithMetrics overrides the default "error not supported" implementation for CreateMetricsReceiver and the default "undefined" stability level.
 func WithMetrics(createMetricsReceiver CreateMetricsFunc, sl component.StabilityLevel) FactoryOption {
-	return factoryOptionFunc(func(o *factory) {
-		o.metricsStabilityLevel = sl
-		o.CreateMetricsFunc = createMetricsReceiver
-	})
+	return internal.WithMetrics(createMetricsReceiver, sl)
 }
 
 // WithLogs overrides the default "error not supported" implementation for CreateLogsReceiver and the default "undefined" stability level.
 func WithLogs(createLogsReceiver CreateLogsFunc, sl component.StabilityLevel) FactoryOption {
-	return factoryOptionFunc(func(o *factory) {
-		o.logsStabilityLevel = sl
-		o.CreateLogsFunc = createLogsReceiver
-	})
+	return internal.WithLogs(createLogsReceiver, sl)
 }
 
 // NewFactory returns a Factory.
 func NewFactory(cfgType component.Type, createDefaultConfig component.CreateDefaultConfigFunc, options ...FactoryOption) Factory {
```

Comment:
Can be 
```suggestion
var NewFactory = internal.NewFactory
```

Reply:
It actually can't. Using a var gives the following error:

```
# go.opentelemetry.io/collector/receiver
./receiver.go:81:18: internal.Internal (type) is not an expression
```

Using a type also doesn't work because the type conversion between the local types and the internal ones doesn't work automagically.